### ISSUE TITLE
~36% of training tokens are never seen, while ~36% are seen twice or more

### ISSUE THREAD

**Issue by sdi1100041** [sandyresearch/parcae#10]
At 16 ranks, ~36% of the corpus is never read by any rank in an epoch, and ~36% of the
tokens that *are* consumed are re-reads of a shard another rank already covered. The
effective unique corpus is ~63% of the nominal one — for us, 21.2B tokens of 33.3B.

The cause is that each rank shuffles the parquet file list with a **different** seed
before taking its stride slice, so `ParquetStreamPure`'s per-rank file sets are
independent random samples rather than a partition. The sibling class in the same file
does the same thing correctly, and all 8 shipped `launch_configs/*.yaml` select the
broken one.

### The two lines

`scripts/train.py:265` gives every rank a different seed:

```python
        seed=(cfg.seed + fabric.global_rank),
```

`recpre/huggingface_dataset.py:160-170` shuffles with that seed, then stride-slices:

```python
        if shuffle_filenames:
            random.Random(seed).shuffle(parquet_files)

        # Shard files for distributed training
        if plan_for_later_rank_expansion_to > 0:
            ranks_for_file_selection = max(plan_for_later_rank_expansion_to, num_processes)
        else:
            ranks_for_file_selection = num_processes
        self.parquet_files = (
            parquet_files[process_rank::ranks_for_file_selection] if ranks_for_file_selection > 1 else parquet_files
        )
```

`list[rank::world]` partitions a list only if every rank slices the **same** list. Here
the ranks slice different permutations, so the slices overlap and leave gaps.

### The same file already does it correctly

`ParquetStream` (`type: pqds`) shuffles *inside* the rank-0 branch and broadcasts the
result, so every rank slices an identical list — a true partition
(`recpre/huggingface_dataset.py:56-79`):

```python
            if process_rank == 0:
                filenames = [str(pth) for pth in sorted(Path(dataset_folder_path).glob(f"{prefix}*"))]
                if shuffle_filenames:
                    random.seed(seed)
                    random.shuffle(filenames)  # inplace
            ...
            torch.distributed.broadcast_object_list(obj, 0)
```

`ParquetStreamPure` (`type: pqds-pure`) broadcasts the **sorted** list first (`:149-159`)
and only then shuffles per-rank — which is what breaks it. Every shipped launch config
uses `pqds-pure`, for both `train_data` and `val_data`.

### Reproduction (stdlib only, no data needed)

```python
import random
N = 576                                   # parquet shards in the corpus

def rank_files(rank, world, base_seed=42, plan=8):
    files = list(range(N))                                # stand-in for sorted() filenames
    random.Random(base_seed + rank).shuffle(files)          # :161, seed = cfg.seed + rank
    return files[rank::max(plan, world)]                   # :165, :169

for world in (8, 16, 32):
    reads = [f for r in range(world) for f in rank_files(r, world)]
    covered = len(set(reads))
    print(f"world={world:3d}  covered {covered}/{N} = {100*covered/N:.1f}%"
          f"  (expected {100*(1-(1-1/world)**world):.1f}%)"
          f"   duplicate reads {len(reads)-covered}/{len(reads)}")
```

```
world=  8  covered 380/576 = 66.0%  (expected 65.6%)   duplicate reads 196/576
world= 16  covered 364/576 = 63.2%  (expected 64.4%)   duplicate reads 212/576
world= 32  covered 364/576 = 63.2%  (expected 63.8%)   duplicate reads 212/576
```

Expected coverage is `1-(1-1/W)^W`, decreasing toward `1-1/e = 63.2%`; individual draws
scatter by about ±1.2 pp, so the numbers above are one draw, not a constant. Replacing the
three lines with a single shared shuffle gives 576/576 with max multiplicity 1.

### Consequence

Token-weighted over 60 base seeds at W=16: **35.6% of tokens never read** (sd 1.4) and
**35.6% of consumed reads duplicated** (sd 1.4). Coverage is size-independent, so the
shard-count and token-count figures agree even though our shard sizes vary a lot
(sd = 33% of the mean).

Any stated token budget or epoch count is therefore wrong in both directions: our
29.6B-token run over a 33.3B-token corpus is ~1.4 epochs over ~21.2B unique tokens, not
the ~0.9 epochs it looks like. And which shards are dropped is a function of `seed`, so
two runs differing only in `seed` train on materially different corpora.

The bug is inert only at `num_processes == 1`, or with `shuffle_filenames: False`
(`recpre/settings.py:191` defaults it to `True`).

### Suggested fix

Make the *filename* shuffle rank-invariant, keeping the per-rank seed for the within-shard
row shuffle:

```diff
         if shuffle_filenames:
-            random.Random(seed).shuffle(parquet_files)
+            # Must not depend on process_rank: the stride slice below partitions
+            # only if every rank shuffles to the same order.
+            random.Random(filename_shuffle_seed).shuffle(parquet_files)
```

with `filename_shuffle_seed` threaded from `cfg.seed` alongside the existing per-rank
`seed` — or simply by moving the shuffle into the rank-0 branch above the broadcast, as
`ParquetStream` already does.

This changes which tokens each rank sees, so runs trained after it are not
data-comparable with runs trained before it.

### Separate problem, not addressed by the above

With unequal shard sizes and `repetitions=None`, the rank holding the least data
exhausts its shards first, and when its iterator ends it leaves the training loop
**without raising** — the remaining ranks then block in the next collective until the
NCCL watchdog fires, while the job reports exit 0. We lost several multi-hour runs this
way. Note this is *not* fixed by the change above: in our corpus shard sizes have
sd = 33% of the mean, so a correct partition still leaves ±25% per-rank token imbalance.
It probably wants its own issue (size-aware assignment, or wraparound, or raising on
exhaustion); happy to file one if useful.

### Environment

Both lines are present in the repository's root commit `62f3055` ("Initial release",
2026-04-15) and are unchanged on `origin/main` today (`6e45195`):
`recpre/huggingface_dataset.py` is byte-identical between the two, and the single
upstream change to `scripts/train.py` since (`a72a482`) does not touch line 265. Our
clone has local modifications to both files, but none on the lines quoted here.

**Comment by Hprairie** [comment]
Thanks for pointing this out, I have added a PR to fix both issues. I verified that the fixes lead to significant speedups and that scaling laws still hold.

Very much the detailed analysis. I have tested locally and think everything is fixed, but let me know if not.

**Comment by sdi1100041** [comment]
Thanks for the quick fix — it resolves both of the issues we reported. However, there is a
third, independent one that this fix does not touch, and we can now say that with the fix in
hand rather than by prediction. See #12
